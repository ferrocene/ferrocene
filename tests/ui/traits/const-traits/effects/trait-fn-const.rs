// Regression test for issue #113378.
#![feature(const_trait_impl, effects)] //~ WARN the feature `effects` is incomplete

#[const_trait]
trait Trait {
    const fn fun(); //~ ERROR functions in traits cannot be declared const
}

impl const Trait for () {
    const fn fun() {} //~ ERROR functions in trait impls cannot be declared const
}

impl Trait for u32 {
    const fn fun() {} //~ ERROR functions in trait impls cannot be declared const
}

trait NonConst {
    const fn fun(); //~ ERROR functions in traits cannot be declared const
}

fn main() {}
