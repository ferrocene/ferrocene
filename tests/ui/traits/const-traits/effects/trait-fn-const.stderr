error[E0379]: functions in traits cannot be declared const
  --> $DIR/trait-fn-const.rs:6:5
   |
LL | #[const_trait]
   | -------------- this declares all associated functions implicitly const
LL | trait Trait {
LL |     const fn fun();
   |     ^^^^^-
   |     |
   |     functions in traits cannot be const
   |     help: remove the `const`

error[E0379]: functions in trait impls cannot be declared const
  --> $DIR/trait-fn-const.rs:10:5
   |
LL | impl const Trait for () {
   |      ----- this declares all associated functions implicitly const
LL |     const fn fun() {}
   |     ^^^^^-
   |     |
   |     functions in trait impls cannot be const
   |     help: remove the `const`

error[E0379]: functions in trait impls cannot be declared const
  --> $DIR/trait-fn-const.rs:14:5
   |
LL |     const fn fun() {}
   |     ^^^^^ functions in trait impls cannot be const
   |
help: remove the `const` ...
   |
LL -     const fn fun() {}
LL +     fn fun() {}
   |
help: ... and declare the impl to be const instead
   |
LL | impl const Trait for u32 {
   |      +++++

error[E0379]: functions in traits cannot be declared const
  --> $DIR/trait-fn-const.rs:18:5
   |
LL |     const fn fun();
   |     ^^^^^ functions in traits cannot be const
   |
help: remove the `const` ...
   |
LL -     const fn fun();
LL +     fn fun();
   |
help: ... and declare the trait to be a `#[const_trait]` instead
   |
LL + #[const_trait]
LL | trait NonConst {
   |

warning: the feature `effects` is incomplete and may not be safe to use and/or cause compiler crashes
  --> $DIR/trait-fn-const.rs:2:30
   |
LL | #![feature(const_trait_impl, effects)]
   |                              ^^^^^^^
   |
   = note: see issue #102090 <https://github.com/rust-lang/rust/issues/102090> for more information
   = note: `#[warn(incomplete_features)]` on by default

error: aborting due to 4 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0379`.
